StorageUtility/Windows/Device.cpp

```diff
@@ -0,0 +1,76 @@
+/*
+<License>
+Copyright 2015 Virtium Technology
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+http ://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+</License>
+*/
+#include "Device.h"
+#include "StorageUtility.h"
+
+namespace vtStor
+{
+
+cDevice::cDevice(const PSP_DEVINFO_DATA& DevInfoData, const PSP_DEVICE_INTERFACE_DATA& DevInterfaceData, const PSP_INTERFACE_DEVICE_DETAIL_DATA& DevDetailData)
+{
+    U32 size = sizeof(DevInfoData->cbSize) + sizeof(DevInfoData->ClassGuid) + sizeof(DevInfoData->DevInst) + sizeof(DevInfoData->Reserved);
+    m_DevInfoData = (PSP_DEVINFO_DATA)LocalAlloc(LPTR, size);
+    m_DevInfoData->cbSize = DevInfoData->cbSize;
+    m_DevInfoData->ClassGuid = DevInfoData->ClassGuid;
+    m_DevInfoData->DevInst = DevInfoData->DevInst;
+    m_DevInfoData->Reserved = DevInfoData->Reserved;
+
+    size = sizeof(DevInterfaceData->cbSize) + sizeof(DevInterfaceData->Flags) + sizeof(DevInterfaceData->InterfaceClassGuid) + sizeof(DevInterfaceData->Reserved);
+    m_DevInterfaceData = (PSP_DEVICE_INTERFACE_DATA)LocalAlloc(LPTR, size);
+    m_DevInterfaceData->cbSize = DevInterfaceData->cbSize;
+    m_DevInterfaceData->Flags = DevInterfaceData->Flags;
+    m_DevInterfaceData->InterfaceClassGuid = DevInterfaceData->InterfaceClassGuid;
+    m_DevInterfaceData->Reserved = DevInterfaceData->Reserved;
+
+    U32 devicePathLength = (std::wcslen(DevDetailData->DevicePath) + 1);
+    size = sizeof(DevDetailData->cbSize) + (devicePathLength * sizeof(DevDetailData->DevicePath));
+    m_DevDetailData = (PSP_INTERFACE_DEVICE_DETAIL_DATA)LocalAlloc(LPTR, size);
+    m_DevDetailData->cbSize = DevDetailData->cbSize;
+    wmemcpy(m_DevDetailData->DevicePath, DevDetailData->DevicePath, devicePathLength);
+}
+
+cDevice::~cDevice()
+{
+    LocalFree( m_DevInfoData );
+    LocalFree( m_DevInterfaceData );
+    LocalFree( m_DevDetailData );
+}
+
+void cDevice::Data(std::unordered_map<eDeviceDataType, void*>& Data)
+{
+    Data.clear();
+    Data.insert({ eDeviceDataType::PSP_DEVINFO_DATA, m_DevInfoData });
+    Data.insert({ eDeviceDataType::PSP_DEVICE_INTERFACE_DATA, m_DevInterfaceData });
+    Data.insert({ eDeviceDataType::PSP_INTERFACE_DEVICE_DETAIL_DATA, m_DevDetailData });
+}
+
+DeviceHandle cDevice::Handle()
+{
+    eErrorCode errorCode;
+    DeviceHandle deviceHandle;
+
+    errorCode = GetStorageDeviceHandle(m_DevDetailData->DevicePath, deviceHandle);
```

Comment:
This function should be remove from the StorageUtility.cpp and move into this class.